### ISSUE TITLE
`@response file` failed to used in clang-tidy.exe with a large project(such as ue4.28)

### ISSUE THREAD

**Issue by SambeauWang** [microsoft/vscode-cpptools#8843]
Bug type: Language Service

**Describe the bug**
- OS and Version: Windows 10 21H2
- VS Code Version: 1.63.1
- C/C++ Extension Version: 1.8.5

**Steps to reproduce**

1. get ue4.28 source code, create default fps demo and build with vs2022
2. use ue4 script to generate vscode workspace
3. Open a cpp file and clang-tidy fail to search include path

**Expected behavior**
(1)  a new choice for use `-p <build-path> ` instead of  `--` for compile info in `clang-tidy`(https://clang.llvm.org/extra/clang-tidy/).
(2) use Clang-tidy feature in unreal engine vscode-cpptool

**Code sample and logs**
- Configurations in `c_cpp_properties.json`
```
{
    "configurations": [
        {
            "name": "ue4_28Editor Editor Win64 Development (ue4_28)",
            "intelliSenseMode": "msvc-x64",
            "compileCommands": "F:\\UE4\\ue4_28\\.vscode\\compileCommands_ue4_28.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        },
        {
            "name": "Win32",
            "intelliSenseMode": "msvc-x64",
            "compileCommands": "F:\\UE4\\ue4_28\\.vscode\\compileCommands_Default.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```

- Logs from running `C/C++: Log Diagnostics` from the VS Code command palette
```
cpptools/getCodeActions: G:\XXX\Source\XXX\Private\XXXEffect.cpp (id: 12)
cpptools/abortRequest
cpptools/getDocumentSymbols
"c:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5/bin/../LLVM/bin/clang-tidy.exe"
--header-filter="(G:/master/project|G:\\master\\project)"
--quiet
"G:/XXX/Source/XXX/Private/XXXEffect.cpp"
--
-std=c++17
-xc++
-Wno-pragma-pack
-includeG:\XXX\Source\Intermediate\Build\Win64\UE4Editor\Development\XXX\Definitions.XXX.h
-include"G:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Development\Engine\SharedPCH.Engine.h"
@"C:/Users/XXX/AppData/Local/Temp/cppADAC.tmp"
Error while processing G:\XXX\Source\XXX\Private\XXXEffect.cpp.
G:\XXX\Source\XXX\Private\XXXEffect.cpp:1:10: error: 'XXXComponent.h' file not found [clang-diagnostic-error]
#include "XXXComponent.h"
         ^~~~~~~~~~~~~~~~~~~~~~~
  Processing folder (recursive): G:/XXX/SOURCE/
  Processing folder (recursive):
```

**Additional context**
I try to use clang-tidy.exe in cmd, and try to use `@response file`. It failed to use `@response file` in clang-tidy.
(1)  I try to use `"C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp -- @G:\master\compile_flags.txt`. The result is ` error: 'XXXComponent.h' file not found`
`@response file` is follewed.
```
...
-IG:\master\ue4_engine\Engine\Source\Runtime\AudioMixerCore 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\PropertyAccess 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess\Public 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\UnrealEd 
...
```

(2) I try to move `--` to `@response file` and merge all args as one line(the command is `"C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp @G:\master\compile_flags.txt`). ` error: 'XXXComponent.h' file not found` is not exit.

`@response file` is follewed.
```
-- ... -IG:\master\ue4_engine\Engine\Source\Runtime\AudioMixerCore -IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\PropertyAccess -IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess\Public -IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess -IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\UnrealEd ...
```

(3) **I try to use `-p XXX.txt` instead of `--`. Such as `"C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp -p G:\master\compile_flags.txt`. It work fine with ue4.28 project.**

`compile_flags.txt` is follewed.
```
...
-IG:\master\ue4_engine\Engine\Source\Runtime\AudioMixerCore 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\PropertyAccess 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess\Public 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\UnrealEd 
...
```

Then I try to use `-p XXX.txt` to `cpptools` extension as args, it fail working. `--` and `-p XXX.txt` may be conflict(https://clang.llvm.org/extra/clang-tidy/).

**Comment by sean-mcmanus** [comment]
Our code analysis implementation doesn't directly process the compile commands file or the response files contained in it -- we rely on the IntelliSense configuration. You should not be trying to pass in any additional "-p or --" as args. Is IntelliSense able to find the headers when opening XXXEffect.cpp? If so, then code analysis should also be able to without any additional configuration.

**Comment by sean-mcmanus** [comment]
FYI, the compile commands generated by Unreal Engine may be incorrect, leading to incorrect IntelliSense and therefore incorrect code analysis configuration.

**Comment by sean-mcmanus** [comment]
Our extension will create a response file when there are too many include paths, which may occur with the Unreal Engine. Are you saying our extension has a bug in that case or is the compile commands itself incorrect?

**Comment by SambeauWang** [comment]
For the log, extension will create a response file with `--` as args when there are too many include paths. There maybe have a some problem which use `--` as clang-tidy compile args will result in `@response file`  failture . So I try to use `clang-tidy.exe` in cmd to test the `response file`.  The command is `"C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp @G:\master\compile_flags.txt`.
```
cpptools/getCodeActions: G:\XXX\Source\XXX\Private\XXXEffect.cpp (id: 12)
cpptools/abortRequest
cpptools/getDocumentSymbols
"c:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5/bin/../LLVM/bin/clang-tidy.exe"
--header-filter="(G:/master/project|G:\\master\\project)"
--quiet
"G:/XXX/Source/XXX/Private/XXXEffect.cpp"
--
-std=c++17
-xc++
-Wno-pragma-pack
-includeG:\XXX\Source\Intermediate\Build\Win64\UE4Editor\Development\XXX\Definitions.XXX.h
-include"G:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Development\Engine\SharedPCH.Engine.h"
@"C:/Users/XXX/AppData/Local/Temp/cppADAC.tmp"
Error while processing G:\XXX\Source\XXX\Private\XXXEffect.cpp.
G:\XXX\Source\XXX\Private\XXXEffect.cpp:1:10: error: 'XXXComponent.h' file not found [clang-diagnostic-error]
#include "XXXComponent.h"
         ^~~~~~~~~~~~~~~~~~~~~~~
  Processing folder (recursive): G:/XXX/SOURCE/
  Processing folder (recursive):
```
`response file` is follewed.
```
...
-IG:\master\ue4_engine\Engine\Source\Runtime\AudioMixerCore 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\PropertyAccess 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess\Public 
-IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess 
-IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\UnrealEd 
...
```
**Then, I try to use `-p` as clang-tidy compile args with same heards info, It work fine with my project.**

**Comment by sean-mcmanus** [comment]
@SambeauWang Are you able to determine why the XXXComponent.h header can't be found, i.e. what folder is it located in and why that folder is not being added as a -I and/or why it's able to be found with -p? Is IntelliSense working with XXXEffect.cpp?

**Comment by SambeauWang** [comment]
@sean-mcmanus I try to use the path of  XXXComponent.h in cmd, such as `C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp -- -IG:/XXX/Source/XXX/Public`. `error: 'XXXComponent.h' file not found` is not exit. 

Then I try to move `--` to `@response file` and merge all args as one line(the command is `"C:\Users\XXX\.vscode\extensions\ms-vscode.cpptools-1.8.5\LLVM\bin\clang-tidy.exe" --quiet XXX.cpp @G:\master\compile_flags.txt)`. error: 'XXXComponent.h' file not found is not exit.
`response file` is follewed.
```
-- ... -IG:\master\ue4_engine\Engine\Source\Runtime\AudioMixerCore -IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\PropertyAccess -IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess\Public -IG:\master\ue4_engine\Engine\Source\Runtime\PropertyAccess -IG:\master\ue4_engine\Engine\Intermediate\Build\Win64\UE4Editor\Inc\UnrealEd ...
```

I think it maybe result in the error which the args is too long in `response file` for `clang-tidy.exe`.

**Comment by Colengms** [comment]
Hi @SambeauWang .  Could this be a duplicate of https://github.com/microsoft/vscode-cpptools/issues/8812 ?  I appear to be some issues with Unreal Editor currently generating incorrect compile_commands files.

**Comment by sean-mcmanus** [comment]
Can you run C/C++: Log Diagnostics after opening XXXEffect.cpp and see whether or not the path to XXXComponent.h is in the Includes list? That should be identical to what we are using in the -I in the response file we generate for clang-tidy.

**Comment by Colengms** [comment]
Hi @SambeauWang.

> (1) a new choice for use -p <build-path> instead of -- for compile info in clang-tidy

We should be passing all of the same arguments as were present in `compile_commands.json` (including response file contents), directly to clang-tidy using `--`.  We support using clang-tidy in scenarios that do not involve compile-commands and are leveraging the results of that process instead of using the existing `compile-commands.json` file in your scenario.  The same information is used to configure IntelliSense.  If clang-tidy is failing to find a header file that would suggest either that same header is missing from IntelliSense, or there is perhaps a bug in which something is wrong or omitted when invoking clang-tidy.

For the source file involved, could you provide the complete build command line?  (The `compile-commands.json` entry for that file, and the contents of the response file it uses.)  If you could also provide the `Log Diagnostics` output Sean requested, we should be able to compare the 2 sets of information to see if there is anything inconsistent between them.

As I mentioned in my previous comment, there are known issues with UE 4.27 that result in headers not being found.  I'm not sure if that has been addressed in 4.28.  Where can I find 4.28?  :)

**Comment by SambeauWang** [comment]
@Colengms Thank you for help. (1) From the log, I think this extension use `compile-commands.json`(including response file contents) to create a temp `response file` which has a lot of `-Ipath` for `clang-tidy.exe`. And This extension use a command, such as `clang.tidy.exe  --quiet XXX.cpp -- @temp response file`. So I try to use same command in cmd, And the result is show above.
(2) There maybe some issues with Unreal Editor currently generating incorrect compile_commands files. So I try to fix it as attachment(change suffix to cs)
[VSCodeProjectFileGenerator.Log](https://github.com/microsoft/vscode-cpptools/files/8075827/VSCodeProjectFileGenerator.Log)
.
(3) UE 4.28 is not release, so I pull master branch.
(4) The log info and rsp file is follewed.
[LogDiagnostics.txt](https://github.com/microsoft/vscode-cpptools/files/8075857/LogDiagnostics.txt)
[XXXPhysicsRuntime.7.txt](https://github.com/microsoft/vscode-cpptools/files/8075858/XXXPhysicsRuntime.7.txt)

**Comment by sean-mcmanus** [comment]
The Log Diagnostics file is actually logging from the C/C++ pane -- can you run the C/C++: Log Diagnostics command and get logging from the C++ Diagnostics pane?

**Comment by SambeauWang** [comment]
The log from the C++ Diagnostics  is follewed.

[C++Diagnostics.txt](https://github.com/microsoft/vscode-cpptools/files/8076126/C%2B%2BDiagnostics.txt)

**Comment by sean-mcmanus** [comment]
It says "No active translation units." -- we need the C/C++: Log Diagnostics command to be run after the repro file (XXXEffect.cpp) is opened and IntelliSense has finished initializing (flame icon goes away).

**Comment by SambeauWang** [comment]
"No active translation units." is still. flame icon goes away, and error is show in `PROBLEMS` panel
[C++Diagnostics2.txt](https://github.com/microsoft/vscode-cpptools/files/8076198/C%2B%2BDiagnostics2.txt)
.

**Comment by SambeauWang** [comment]
![image](https://user-images.githubusercontent.com/22713858/154184459-331851cf-9245-4b8d-aa3f-6cf587d1178b.png)

**Comment by sean-mcmanus** [comment]
What is the error in the problems window? Does it say it can't find the header XXXComponent.h? Can you make sure C_Cpp.intelliSenseEngineFallback is set to "Disabled"?

**Comment by SambeauWang** [comment]
![image](https://user-images.githubusercontent.com/22713858/154186173-fb6bcd1a-c510-4f58-a7ee-12d7d58e027c.png)
![image](https://user-images.githubusercontent.com/22713858/154186259-76ef2edf-074a-4016-889f-44b1665e0996.png)

**Comment by sean-mcmanus** [comment]
Those are clang diagnostic errors -- are you getting any IntelliSense errors (they appear as `C++(<errorCodeNumber>)`). Something is causing the IntelliSense process to not startup correctly. Does your "C/C++" logging (not the "diagnostics" one) show any "crash" messages? Your C_Cpp.intelliSenseEngine is set to "Default", right?

**Comment by SambeauWang** [comment]
C_Cpp.intelliSenseEngine is set to "Tag parse". "C/C++" logging([LogDiagnostics.txt](https://github.com/microsoft/vscode-cpptools/files/8075857/LogDiagnostics.txt)) is showed above.

**Comment by SambeauWang** [comment]
After C_Cpp.intelliSenseEngine is set to "Default", the log of `C/C++: Log Diagnostics `:
[C++Diagnostics3.txt](https://github.com/microsoft/vscode-cpptools/files/8076399/C%2B%2BDiagnostics3.txt)

**Comment by SambeauWang** [comment]
@sean-mcmanus  Thank you for help. Now I use VS2022 to work with Unreal Engine. But VSCode and cpptool extension is more lightweight and friendly. So I try to set C_Cpp.intelliSenseEngine is set to "Tag parse", and want to use clang-tidy.exe to do a simple grammar check.

It maybe result in `it seems like they may have been generated using different source files` which I replace some name with `XXX`. In `error: 'XXXVehicleComponent.h' file not found`, ``XXXVehicleComponent.h` is location in `"G:\master\XXX_project\Plugins\XXXPhysics\Source\XXXPhysicsRuntime\Public"`. And there is no errors in IntelliSense. The path of 'XXXVehicleComponent.h' header file is not missing for `XXXPhysicsRuntime.7.rsp` and ``,  and I think that it is more possibility that `clang-tidy` is conflict with `@response file` or args is too long for `clang-tidy.exe`. Because  I try to use same command in cmd from `LogDiagnostics.txt`, the result show above.

![image](https://user-images.githubusercontent.com/22713858/154384278-59e6c97c-1cf7-4d7c-b72d-b86ea480cd77.png)

**Comment by sean-mcmanus** [comment]
FYI, VS also has some clang-tidy support, but it's somewhat different: https://docs.microsoft.com/en-us/cpp/code-quality/clang-tidy?view=msvc-170 .

I've filed a feature request on enabling clang-tidy without IntelliSense at https://github.com/microsoft/vscode-cpptools/issues/8874 .

The `G:\master\ue4_engine\Engine\Source\Runtime\Core\Public` was not listed in the Includes for IntelliSense. If that was supposed to be then there could be some bug in our processing of the compile_commands.json/rsp file.

**Comment by sean-mcmanus** [comment]
Yeah, I'm seeing @response files not working with clang-tidy -- seems like our tests were inheriting the include paths from the environment instead of getting them from the response file :(

**Comment by sean-mcmanus** [comment]
This should be fixed with 1.10.2 (pre-release).
